---
title: "Demonstration of HyperTraPS in R"
output: html_document
---

If we just want the HyperTraPS code without other dependencies, we only need Rcpp, and can use the following

```{r}
library(Rcpp)
sourceCpp("hypertraps-r.cpp")
```

If we want various helper functions and ggplot functions in addition, use this

```{r}
source("hypertraps.R")
```

### Simple demo 

Construct simple synthetic dataset, including timings, for the "cross" double-pathway system

```{r}
m.1 = matrix(rep(c(0,0,0,0,0,
               1,0,0,0,0,
               1,1,0,0,0,
               1,1,1,0,0,
               1,1,1,1,0,
               0,0,0,0,0,
               0,0,0,0,1,
               0,0,0,1,1,
               0,0,1,1,1,
               0,1,1,1,1),5), byrow=TRUE, ncol=5)
m.2 = matrix(rep(c(1,0,0,0,0,
               1,1,0,0,0,
               1,1,1,0,0,
               1,1,1,1,0,
               1,1,1,1,1,
               0,0,0,0,1,
               0,0,0,1,1,
               0,0,1,1,1,
               0,1,1,1,1,
               1,1,1,1,1),5), byrow=TRUE, ncol=5)
times = rep(c(0.1, 0.2, 0.3, 0.4, 0.5), 10)
```

Run HyperTraPS with start and end observations, start and end timings, and feature labels

```{r}
my.post = HyperTraPS(m.2, initialstates_arg = m.1, 
                     starttimes_arg = times, endtimes_arg = times,
                     featurenames_arg = c("A", "B", "C", "D", "E")) 
```

"Bubble" plot summary. The size of a circle at point x,y gives the probability that feature y is acquired at ordinal time x.

```{r}
plotHypercube.bubbles(my.post)
```

Sampled routes on the hypercubic space. Each edge is a transition, corresponding to the acquisition of a new feature. The edges are labelled by the feature acquired and the statistics of the time taken (mean +- s.d.). The edge width gives the probability flux through that transition.

```{r}
plotHypercube.sampledgraph2(my.post)
```

### Input/output between R data structure and file format

Write output to files

```{r}
writeHyperinf(my.post, "simpledemo", my.post$L, postlabel = "simpledemo", fulloutput=TRUE)
```

Retrieve output from files

```{r}
my.post.r = readHyperinf("simpledemo", postlabel = "simpledemo", fulloutput=TRUE)
```

### Other visualisations

Motif plot, containing the same data as the bubble plots but via differen visualisation

```{r}
plotHypercube.motifs(my.post)
```

Time series plot. Each step up the vertical axis corresponds to the acquisition of a new feature. The horizontal axis gives the time of the corresponding event, and the colour labels the feature that was acquired.

```{r}
plotHypercube.timeseries(my.post)
```

### Other functional examples

Run an example with fewer walkers

```{r}
my.post.sparse = HyperTraPS(m.2, initialstates_arg = m.1, 
                            starttimes_arg = times, featurenames_arg = c("A", "B", "C", "D", "E"), 
                            walkers_arg = 2,
                            limited_output_arg = 1)
```

Stepwise regularisation

```{r}
my.post.regularise = HyperTraPS(m.2, initialstates_arg = m.1, regularise_arg = 1,
                                walkers_arg = 20,
                                limited_output_arg = 1)
plotHypercube.regularisation(my.post.regularise)
plotHypercube.summary(my.post.regularise)
```

Use simulated annealing to get a maximum likelihood estimate rather than a Bayesian picture

```{r}
my.post.sa = HyperTraPS(m.2, initialstates_arg = m.1, sa_arg = 1,
                        limited_output_arg = 1)
plotHypercube.summary(my.post.sa)
```

Use "phenotype landscape inference" -- unbiased sampling. This takes longer but is more flexible with uncertain data

```{r}
my.post.pli = HyperTraPS(m.2, initialstates_arg = m.1, PLI_arg = 1,
                         limited_output_arg = 1)
plotHypercube.summary(my.post.pli)
```

HyperTraPS supports different parameter structures. This approach allows every edge on the hypercube to have its own independent parameters, then regularises to remove extraneous parameters

```{r}
my.post.bigmodel.regularise = HyperTraPS(m.2, initialstates_arg = m.1, model_arg = -1,
                                         regularise_arg = 1, walkers_arg = 20,
                                         limited_output_arg = 1)
plotHypercube.regularisation(my.post.bigmodel.regularise)
```

### Scientific examples

A set of short-form examples from past studies -- should run in a few minutes and give approximations to the original results

1. Ovarian cancer case study: traits are chromosomal aberrations, observations are independent patient samples

```{r}
cgh.mat = readLines("RawData/ovarian.txt")
cgh.mat = do.call(rbind, lapply(strsplit(cgh.mat, ""), as.numeric))
cgh.names = as.vector(read.table("RawData/ovarian-names.txt", sep=","))[[1]]

my.post.cgh = HyperTraPS(cgh.mat, 
                        length_index_arg = 3, outputinput_arg = 1, 
                        featurenames_arg = cgh.names,
                        limited_output_arg = 1) 
ggarrange(plotHypercube.lik.trace(my.post.cgh), 
          plotHypercube.bubbles(my.post.cgh, reorder=TRUE), 
          plotHypercube.sampledgraph2(my.post.cgh, no.times=TRUE), nrow=3)
plotHypercube.sampledgraph2(my.post.cgh, no.times=TRUE)
```

2. C4 photosynthesis case study: traits are physical/genetic features associated with C4, observations are (incomplete) phylogenetically independent intermediate species

```{r}
c4.mat = as.matrix(read.table("RawData/c4-curated.csv", sep=","))
c4.names = as.vector(read.table("RawData/c4-trait-names.txt", sep=","))[[1]]

my.post.c4 = HyperTraPS(c4.mat, 
                        length_index_arg = 3, 
                        losses_arg = 1,
                        featurenames_arg = c4.names,
                        limited_output_arg = 1) 
ggarrange(plotHypercube.lik.trace(my.post.c4), plotHypercube.bubbles(my.post.c4, reorder=TRUE), nrow=2)
```

3. Severe malaria disease progression case study: traits are clinical features, observations are (incomplete) independent patient presentations

```{r}
malaria.df = read.csv("RawData/jallow_dataset_binary_with2s.csv")
malaria.mat = as.matrix(malaria.df[,2:ncol(malaria.df)])
malaria.names = as.vector(read.table("RawData/malaria-names.txt", sep=","))[[1]]

my.post.malaria = HyperTraPS(malaria.mat, 
                        length_index_arg = 3,
                        kernel_index_arg = 2,
                        walkers_arg = 20,
                        featurenames_arg = malaria.names,
                        limited_output_arg = 1) 
ggarrange(plotHypercube.lik.trace(my.post.malaria), plotHypercube.bubbles(my.post.malaria, reorder=TRUE, transpose=TRUE), nrow=2)
```

4. Tool use evolution case study: traits are modes of tool use, observations are phylogenetically coupled species observations (phylogeny has been accounted for, giving transition pairs)

```{r}
tools.mat = as.matrix(read.table("RawData/total-observations.txt-trans.txt"))
tools.names = as.vector(read.table("RawData/tools-names.txt"))[[1]]
tools.starts = tools.mat[seq(from=1, to=nrow(tools.mat), by=2),]
tools.ends = tools.mat[seq(from=2, to=nrow(tools.mat), by=2),]

my.post.tools = HyperTraPS(tools.ends, initialstates_arg = tools.starts, 
                           length_index_arg = 3, 
                           featurenames_arg = tools.names,
                           limited_output_arg = 1) 
ggarrange(plotHypercube.lik.trace(my.post.tools), plotHypercube.bubbles(my.post.tools, reorder=TRUE, transpose=TRUE), nrow=2)
plotHypercube.sampledgraph2(my.post.tools, node.labels = FALSE, max=100)
```

Plot summaries of all scientific case studies
```{r}
ggarrange( plotHypercube.bubbles(my.post.cgh, reorder=TRUE) + xlab("Order") + ylab("Aberration") + ggtitle("Ovarian cancer progression"),
           plotHypercube.bubbles(my.post.c4, reorder=TRUE) + xlab("Order") + ylab("C4 feature") + ggtitle("C4 photosynthesis"),
           plotHypercube.bubbles(my.post.malaria, reorder=TRUE, transpose=TRUE) + xlab("Symptom") + ylab("Order") + ggtitle("Severe malaria progression"),
           plotHypercube.bubbles(my.post.tools, reorder=TRUE, transpose=TRUE) + xlab("Tool use mode") + ylab("Order") + ggtitle("Tool use emergence"))
```

